docs/_bin/generate-license.py

```diff
@@ -0,0 +1,696 @@
+    if license['license_name'] != 'Apache License version 2.0':
+        license_phrase += " which is available under {}".format(license['license_name'])
+    if is_non_empty(license, 'additional_license_statement'):
+        license_phrase += ", {}".format(license['additional_license_statement'])
+    if is_non_empty(license, 'license_file_path'):
+        license_file_list = []
+        if type(license['license_file_path']) == list:
+            license_file_list.extend(license['license_file_path'])
+        else:
+            license_file_list.append(license['license_file_path'])
+        if len(license_file_list) == 1:
+            license_phrase += ". For details, see {}".format(license_file_list[0])
+        else:
+            license_phrase += ". For details, "
+            for each_file in license_file_list:
+                if each_file == license_file_list[-1]:
+                    license_phrase += ", and {}".format(each_file)
+                elif each_file == license_file_list[0]:
+                    license_phrase += "see {}".format(each_file)
+                else:
+                    license_phrase += ", {}".format(each_file)
+    
+    license_phrase += "."
+
+    print_license_phrase(license_phrase)
+
+    if 'source_paths' in license:
+        for source_path in license['source_paths']:
+            if type(source_path) is dict:
+                for class_name, path in source_path.items():
+                    print("      {}:".format(class_name))
+                    print("      * {}".format(path))
+            else:
+                print("      * {}".format(source_path))
+
+    if 'libraries' in license:
+        for library in license['libraries']:
+            if type(library) is not dict:
+                raise Exception("Expected dict but got {}[{}]".format(type(library), library))
+            if len(library) > 1:
+                raise Exception("Expected 1 groupId and artifactId, but got [{}]".format(library))
+            for group_id, artifact_id in library.items():
+                print("      * {}:{}".format(group_id, artifact_id))
+
+
+def check_licenses(license_yaml, dependencie_reports_root):
+    reported_dep_to_licenses = {}
+    compatible_license_names = build_compatible_license_names()
+    for dirpath, dirnames, filenames in os.walk(dependencie_reports_root):
+        for filename in filenames:
+            if filename == "dependencies.html":
+                full_path = os.path.join(dirpath, filename)
+                print_error("Parsing {}".format(full_path))
+                with open(full_path) as report_file:
+                    parser = DependencyReportParser(compatible_license_names)
+                    reported_dep_to_licenses.update(parser.parse(report_file))
+    
+    print_error("Found {} reported licenses".format(len(reported_dep_to_licenses)))
+
+    # Compare licenses in registry and those in dependency reports
```

Comment:
could you add a few additional comments about what is going on here to break up all these loops? makes it kind of rough to follow along with what is going on here

Reply:
Added more comments here and other places. Please let me know if it's still unclear.